docs/exernal-services.md

```diff
@@ -0,0 +1,206 @@
+
+## ExternalReport
+Has a `default_report_for_source_type`. When an ExternalActivity has a tool with this source_type is assigned to a class, then report buttons for this assignment will automatically be added.
+
+ExternalActivities can also have 1 or more ExternalReports directly associated with them. This makes it possible for an author or admin to setup a special report or dashboard for a specific resource.
+
+## ExternalActivity
+Authors and admins can duplicate ExternalActivities. If the ExternalActivity#tool.source_type is "LARA" then the a special `duplicate_on_lara` function is called. This sends a http request to `#{ExternalActivity#url}/remote_duplicate`. It expects LARA to copy the sequence or activity and then return the same kind of data that would normally be sent when an author "publishes" the sequence or activity to the Portal. Then the Portal does this publishing internally using the `ActivityRuntimeAPI`. This internal publishing happens as a 'update' operation not a 'create' operation. This is because the ExternalActivity already exists.
+
+When the ActivityRuntimeAPI is used to create or update an ExternalActivity, the `launch_url` is only set when it is created. Because of this the duplicate_on_lara code sets the `launch_url` first before doing the internal publish.
+
+When data_helpers.rb generates JSON from the activity it includes lara_activity_or_sequence which is based on whether the activity has `launch_url`. This seems be used to determine which links are shown to the user.
+
+## Client
+Clients are used for authentication. When the Portal creates an AccessGrant to provide a token to an external service, it generally has an associated Client. The Client is used to verify requests with the token. If the client specifies domain_matchers then only requests from those domains are accepted.
+
+When the request is part of an OAuth2 authorize, then the Client needs to specify the exact redirect that the Portal is sending this token or code back to. These OAuth2 authorize requests are used when LARA is doing OAuth2. They are also used by the single page apps that use the Portal for authentication like the token-service, glossary authoring, or a portal-report when not launched directly from the portal.
+
+## Seeding the Database
+To make it easier for developers, when the database is seeded a few of the resources above are automatically created.
+
+An ExternalReport is created for the portal-report. And a Client is created that this ExternalReport uses.
+
+# Ways to Improve
+
+## Eliminate the source_type
+
+### default_report
+The use of source_type to connect ExternalActivities to default reports is not needed. Instead a 'default_report' can be added to the Tool model. This will have the same behavior and simplify the modeling.
```

Comment:
```suggestion
The use of `source_type` to connect ExternalActivities to default reports is not needed. Instead a `default_report` can be added to the Tool model. This will have the same behavior and simplify the modeling.
```